pallets/contracts/src/lib.rs

```diff
@@ -0,0 +1,353 @@
+// This file is part of the Polymesh distribution (https://github.com/PolymathNetwork/Polymesh).
+// Copyright (c) 2020 Polymath
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, version 3.
+
+// This program is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::Encode;
+use frame_support::{
+    decl_error, decl_event, decl_module, decl_storage,
+    dispatch::{DispatchError, DispatchResult, DispatchResultWithPostInfo},
+    ensure,
+    traits::Get
+};
+use frame_system::{self as system, ensure_signed};
+use pallet_contracts::{BalanceOf, CodeHash, ContractAddressFor, Gas, Schedule};
+use pallet_identity as identity;
+use polymesh_common_utilities::{
+    identity::Trait as IdentityTrait,
+    protocol_fee::{ChargeProtocolFee, ProtocolOp},
+    Context, with_transaction
+};
+use polymesh_primitives::{IdentityId, SmartExtensionMetadata, TemplateMetadata};
+use sp_core::crypto::UncheckedFrom;
+use sp_runtime::{
+    traits::{Hash, Saturating, StaticLookup},
+    Perbill, SaturatedConversion,
+};
+use sp_std::{marker::PhantomData, prelude::*};
+
+type Identity<T> = identity::Module<T>;
+
+/// Nonce based contract address determiner.
+///
+/// Address calculated from the code (of the constructor), input data to the constructor,
+/// the account id that requested the account creation and the latest nonce of the account id.
+///
+/// Formula: `blake2_256(blake2_256(code) + blake2_256(data) + origin + blake2_256(nonce))`
+pub struct NonceBasedAddressDeterminer<T: Trait>(PhantomData<T>);
+impl<T: Trait> ContractAddressFor<CodeHash<T>, T::AccountId> for NonceBasedAddressDeterminer<T>
+where
+    T::AccountId: UncheckedFrom<T::Hash> + AsRef<[u8]>,
+{
+    fn contract_address_for(
+        code_hash: &CodeHash<T>,
+        data: &[u8],
+        origin: &T::AccountId,
+    ) -> T::AccountId {
+        let data_hash = T::Hashing::hash(data);
+        let nonce = ExtensionNonce::mutate(|n| {
+            *n = *n + 1u64;
+            *n
+        });
+        let nonce_hash = T::Hashing::hash(&(nonce.encode()));
+        let mut buf = Vec::new();
```

Comment:
Optimization: use `Vec::with_capacity(sum_len)` where `sum_len` is computed by summing the length of all slices you're `.extend_from_slice`ing with.